File: tools/packager/merkle.py
"""Merkle tree matching ota-common/src/merkle.cpp byte-for-byte.

Conventions:
  - leaf[i]  = SHA-256(block_i)
  - internal = SHA-256(left || right)
  - odd level: the unpaired trailing node is promoted unchanged.
  - a proof lists only real siblings, leaf -> root; promotion contributes none.
"""

from __future__ import annotations

import hashlib
from typing import List


def hash_leaf(block: bytes) -> bytes:
    return hashlib.sha256(block).digest()


def hash_pair(left: bytes, right: bytes) -> bytes:
    return hashlib.sha256(left + right).digest()


def _reduce(level: List[bytes]) -> List[bytes]:
    out: List[bytes] = []
    i = 0
    while i + 1 < len(level):
        out.append(hash_pair(level[i], level[i + 1]))
        i += 2
    if i < len(level):
        out.append(level[i])  # promote odd tail
    return out


def leaves_for(payload: bytes, block_size: int) -> List[bytes]:
    leaves = []
    for off in range(0, len(payload), block_size):
        leaves.append(hash_leaf(payload[off : off + block_size]))
    return leaves


def merkle_root(leaves: List[bytes]) -> bytes:
    if not leaves:
        return b"\x00" * 32
    level = list(leaves)
    while len(level) > 1:
        level = _reduce(level)
    return level[0]


def merkle_proof(leaves: List[bytes], index: int) -> List[bytes]:
    proof: List[bytes] = []
    if index >= len(leaves):
        return proof
    level = list(leaves)
    idx = index
    while len(level) > 1:
        is_last_odd = (idx == len(level) - 1) and (len(level) % 2 == 1)
        if not is_last_odd:
            sib = idx + 1 if idx % 2 == 0 else idx - 1
            proof.append(level[sib])
        idx //= 2
        level = _reduce(level)
    return proof


def merkle_verify(leaf: bytes, index: int, leaf_count: int, proof: List[bytes], root: bytes) -> bool:
    if leaf_count == 0 or index >= leaf_count:
        return False
    h = leaf
    idx = index
    count = leaf_count
    p = 0
    while count > 1:
        is_last_odd = (idx == count - 1) and (count % 2 == 1)
        if not is_last_odd:
            if p >= len(proof):
                return False
            sib = proof[p]
            p += 1
            h = hash_pair(h, sib) if idx % 2 == 0 else hash_pair(sib, h)
        idx //= 2
        count = (count + 1) // 2
    return p == len(proof) and h == root

File: src/merkle.cpp
#include "ota_common/merkle.h"

#include <cstring>

namespace ota_common {

bool Hash256::operator==(const Hash256 &o) const {
  return std::memcmp(bytes, o.bytes, kSha256Len) == 0;
}

Hash256 hash_leaf(const uint8_t *block, size_t len) {
  Hash256 h;
  Sha256::hash(block, len, h.bytes);
  return h;
}

Hash256 hash_pair(const Hash256 &left, const Hash256 &right) {
  Sha256 s;
  s.update(left.bytes, kSha256Len);
  s.update(right.bytes, kSha256Len);
  Hash256 h;
  s.finish(h.bytes);
  return h;
}

// Reduce one level, promoting an unpaired trailing node.
static std::vector<Hash256> reduce_level(const std::vector<Hash256> &level) {
  std::vector<Hash256> next;
  next.reserve((level.size() + 1) / 2);
  size_t i = 0;
  for (; i + 1 < level.size(); i += 2) {
    next.push_back(hash_pair(level[i], level[i + 1]));
  }
  if (i < level.size())
    next.push_back(level[i]); // promote odd tail
  return next;
}

Hash256 merkle_root(const std::vector<Hash256> &leaves) {
  if (leaves.empty()) {
    Hash256 zero;
    std::memset(zero.bytes, 0, kSha256Len);
    return zero;
  }
  std::vector<Hash256> level = leaves;
  while (level.size() > 1)
    level = reduce_level(level);
  return level[0];
}

std::vector<Hash256> merkle_proof(const std::vector<Hash256> &leaves,
                                  size_t index) {
  std::vector<Hash256> proof;
  if (index >= leaves.size())
    return proof;

  std::vector<Hash256> level = leaves;
  size_t idx = index;
  while (level.size() > 1) {
    const bool idx_is_last_odd =
        (idx == level.size() - 1) && (level.size() % 2 == 1);
    if (!idx_is_last_odd) {
      const size_t sib = (idx % 2 == 0) ? idx + 1 : idx - 1;
      proof.push_back(level[sib]);
    }
    idx /= 2;
    level = reduce_level(level);
  }
  return proof;
}

bool merkle_verify(const Hash256 &leaf, size_t index, size_t leaf_count,
                   const std::vector<Hash256> &proof, const Hash256 &root) {
  if (leaf_count == 0 || index >= leaf_count)
    return false;

  Hash256 h = leaf;
  size_t idx = index;
  size_t count = leaf_count;
  size_t p = 0;
  while (count > 1) {
    const bool idx_is_last_odd = (idx == count - 1) && (count % 2 == 1);
    if (!idx_is_last_odd) {
      if (p >= proof.size())
        return false;
      const Hash256 &sib = proof[p++];
      h = (idx % 2 == 0) ? hash_pair(h, sib) : hash_pair(sib, h);
    }
    idx /= 2;
    count = (count + 1) / 2;
  }
  return p == proof.size() && h == root;
}

bool merkle_verify_block(const uint8_t *block, size_t block_len, size_t index,
                         size_t leaf_count, const std::vector<Hash256> &proof,
                         const Hash256 &root) {
  return merkle_verify(hash_leaf(block, block_len), index, leaf_count, proof,
                       root);
}

} // namespace ota_common
